### ISSUE TITLE
Support compressed send (`-c`)

### ISSUE THREAD

**Issue by johnramsden** [zrepl/zrepl#182]
I imagine this is somewhere on the road map, but I couldn't find it in the projects page, or any feature requests. I did see it was mentioned on the advanced features, we couldn't find reference to it in any current work.

I think it would make sense to transparently use compressed send if the version of ZFS supports it. It should just entail adding the '-c' flag if the ZFS version is high enough.

If this feature isn't yet in the process of being worked on, and if I could be pointed in the right direction, in terms of where this flag should be added, I would be happy to contribute to the feature.

**Comment by problame** [comment]
The main issue with compressed send (as with other newer send flags) is that it requires support on the receiving system as well.
My plan was to have some sort of feature negotiation protocol (think `Accept` header in HTTP).

From the ZoL man page:

```
    -e, --embed
           Generate a more compact stream by using WRITE_EMBEDDED records for blocks which are stored more compactly on disk by the embedded_data pool feature.  This flag has no effect if the embedded_data feature is disabled.  The receiving system
           must have the embedded_data feature enabled.  If the lz4_compress feature is active on the sending system, then the receiving system must have that feature enabled as well. Datasets that are sent with this flag may not be received as an
           encrypted dataset, since encrypted datasets cannot use the embedded_data feature.  See zpool-features(5) for details on ZFS feature flags and the embedded_data feature.

       -c, --compressed
           Generate a more compact stream by using compressed WRITE records for blocks which are compressed on disk and in memory (see the compression property for details).  If the lz4_compress feature is active on the sending system, then the
           receiving system must have that feature enabled as well.  If the large_blocks feature is enabled on the sending system but the -L option is not supplied in conjunction with -c, then the data will be decompressed before sending so it can
           be split into smaller block sizes.

       -w, --raw
           For encrypted datasets, send data exactly as it exists on disk. This allows backups to be taken even if encryption keys are not currently loaded. The backup may then be received on an untrusted machine since that machine will not have
           the encryption keys to read the protected data or alter it without being detected. Upon being received, the dataset will have the same encryption keys as it did on the send side, although the keylocation property will be defaulted to
           prompt if not otherwise provided. For unencrypted datasets, this flag will be equivalent to -Lec.  Note that if you do not use this flag for sending encrypted datasets, data will be sent unencrypted and may be re-encrypted with a differ‐
           ent encryption key on the receiving system, which will disable the ability to do a raw send to that system for incrementals.

```

A recent [Twitter poll](https://twitter.com/problame/status/1133121507532660736) suggests that many people want per-dataset knobs for requiring / disabling encrypted send & recv.

I wonder whether the same applies to compressed send & recv: would the user be OK with automatic fallback to uncompressed send & recv if the receiving system doesn't support the feature?

With regard to your contribution, I think it would be most useful to draft the user interface to the compressed send feature, i.e. **designing an extension of the configuration format**.

Think about the following scenarios:
* receiving system supports compressed recv 
* receiving system supports compressed recv but admin policy prohibits that (for whatever reason)
* receiving system ZFS version does not support compressed recv
* sending job doesn't know the above in advance / above admin preferences might change between send attempts
    * => the receiving-side preferences need to be discovered dynamically

Please provide a draft of job configurations on sending & receiving side that address above scenarios in a simple way. Feel free to invent new semantics, etc.

**Comment by johnramsden** [comment]
@problame Could you elaborate on what you meant by _"receiving system supports compressed recv but admin policy prohibits that (for whatever reason)"_? Do you mean some sort of configuration in zrepl to specify the admin doesn't want to allow compressed receive? Or do you mean some sort of system wide setting?

**Comment by problame** [comment]
There's no system-wide setting nor a zfs recv option that I'm aware of to administratively prohibit recv of compressed streams, hence I meant a zrepl receive side configuration option.
The idea was that you might not want to receive the data compressed because of decompression overhead. But then again, I'm not aware of a compression algorithm in ZFS that has significant *de*compression overhead. And, given the ongoing discussion of enabling compression by default in ZFS, I see little reason to keep that requirement.

**Comment by siilike** [comment]
Note that compressed send is not always compressed and there are cases where the data could be recompressed for a better ratio (reasonable for backups, but not for normal use), so it makes sense to also have zrepl-level compression, zstd would be a good choice.

In that case the level of compression needs negotiating: have a maximum on both sides (for compressing only) and the requestor sets their preference (which might or might not default to the maximum available).

**Comment by problame** [comment]
Would we like transport-level ('zrepl-level') compression configurable per dataset or per job?

I would base that decision on whether zstd gives up early (i.e. doesn't waste CPU time) if the data is incompressible.

Anyways, I have a draft config syntax for send features, will need to play around with it for a few more days though.

**Comment by siilike** [comment]
Why not have both? There are use cases where it makes sense to tune the compression level per dataset. Per job configuration would then be just the default value.

**Comment by problame** [comment]
> Why not have both?
Because it clutters the config format & parsing code significantly, and users will likely want matching syntax (like with the `filesystems` zrepl job config attribute), precendence / inheritance functionality, etc.

And curiously, ZFS already supports all of that with property inheritance, so we could consider just using ZFS properties for configuring send options. 
But then again, how do we handle settings that are different for multiple push jobs that push the same filesystem...

**Comment by siilike** [comment]
Well, why not like this:
```
jobs:
- type: source
  compression: { level: 5 }, # default
  filesystems: {
    "tank<": true,
    "tank/foo<": false, 
    "tank/foo<": { include: false }, # equivalent of the above 
    "tank/foo/bar": { compressionLevel: 9 },
  }
```
The user can specify a per-job default that applies to all datasets and per-dataset options should they want it.

I cannot see any clutter for the feature set it gives.

**Comment by johnramsden** [comment]
Well I was thinking about configuration, I have a few thoughts, and questioned whether adding new configuration options was necessary at all.

If you need to check with the source and destination first anyway to see if compressed send is supported, why not:

* Check if support is there on both ends
* If it is, and the source is compressed, used compressed send
* Otherwise regular send used

Do you really need to add another configuration option? What benefit would there be by disabling compressed send if support existed?

You mentioned wanting the option to disable compressed send to avoid the decompression overhead, but when compressed send is used, the receiving side dataset doesn't decompress the data before storing it, it just stores the already compressed dataset. So I don't see a benefit of being able to disable it.

I question how much benefit you would get from using another source for compression like zstd. Yes you could get better compression, but it will mean another dependency, as well as re-compressing the dataset. If the dataset was compressed already, and had a low compression ratio, effects of zstd would likely not be that significant, since your data must not have been very compressible. I guess where might make a difference would be on very low-bandwidth networks.

The benefits of using the built-in compressed send as opposed to using another tool all together would be since the data is already compressed, you don't have to spend the CPU time to re-compress before sending the data.

**Comment by siilike** [comment]
There can be a significant difference.

Firstly, ZFS zend will send compressed data only when it is already compressed.

Secondly, the compression ratio largely depends on the recordsize parameter. If the application requires a small recordsize that does not compress well does not mean we cannot use a bigger one for transport, the difference can be several times, with the same algorithm.

Thirdly, the algorithms and compression levels can be different. You will receive a better compression ratio with gzip than with lz, even if the data is already compressed with lz, although the result will be better on raw data.

So, there are different scenarios for the sender:
  - data not compressed by ZFS, will be compressed in transport;
  - data compressed by ZFS, will not be compressed in transport;
  - data compressed by ZFS, will be compressed in transport;
  - data compressed by ZFS, will be decompressed and compressed in transport.

The results depend significantly on the nature of the data, sometimes it is worth it, sometimes not. But it does not affect only "very low-bandwidth networks", bandwith costs, needs to be used for other things and there can be time constraints.

It is also reasonable to expect that on a backup system the recordsize parameter would be higher to achieve a better compression ratio, in those cases the data would have to be recompressed anyway, perhaps also with a different algorithm. This is a very important aspect that should also be considered, although not related to compression in transport.

**Comment by johnramsden** [comment]
I see, maybe using an external compression method like zstd should be considered a different case with different configuration options then. 

* Both sides supports compressed send, and data compressed on source, use compressed send always.
* User specifies `compression: <type>`.

Other question is, if the source isn't compressed, should the default be to use an external compression type unless the user specifies `compression: false`?

**Comment by problame** [comment]
At first, sorry for the late reply.
Thanks for your very good and detailed analysis, in particular for pointing out all the different combinations that we need to consider!

I have created #197 to deal with the recordsize & REcompression use-case.
Would you mind investigating whether recompression is possible if the send stream was created with `-c`? Please report back there.

With regards to user-facing configuration: yes, I'm absolutely for 'smart' defaults, i.e. automatically use whatever is available.
With regards to when to apply transport-level compression: **there is no good heuristic** because, AFAIK, there is no way to determine for a dataset what percentage of blocks is _actually_ compressed with the currently configured compression algorithm (Is that true?).
Thus, I would opt to default to always enable transport-level compression, and allow for some way to disable it (either via config or a ZFS property).

To keep the replication protocol extensible, we'd likely also need negotiation of compression algorithms.

**Comment by mdtancsa** [comment]
Would love to see transport level compression independent of what zfs can / cant do. Before I started using zrepl, I would pipe the zfs send through pigz and it would either have very little impact / overhead on data that could not be compressed (.mp4s), or would be a large gain on some of my servers which have a lot of verbose .xml files.

**Comment by Thoro** [comment]
Could we just allow the -c option, or generally allow to pass options to zfs send based on user responsibility?

I have a dataset with 100 TB compressdata, which is uncompressed around 250 TB and I would rather not resend all of that data when I can just send 100 TB - currently the only reason not to use zrepl.

FYI to also contribute to the discussion, if the kernel option `zfs_compressed_arc_enabled=0` is set zfs send always sends uncompressed.

**Comment by problame** [comment]
> Could we just allow the -c option, or generally allow to pass options to zfs send based on user responsibility?
You can do this right now by inserting a `zfs` wrapper script in the `PATH` (how depends on your init system).

Example: https://github.com/zrepl/zrepl/blob/master/platformtest/logmockzfs/zfs

**Comment by Thoro** [comment]
ah great, thanks

**Comment by klauspost** [comment]
If there is anything I can do to assist with this feel free to ping me. Obv. author of https://github.com/klauspost/compress

I wouldn't worry to much about recompressing already compressed data. Most Go compression implementations handle this very well. zstd (mine) skips at about 2-3GB/s, lz4/s2/gzip (mine) at ~4GB/s, so likely it should be able to keep up with IO in these cases.

**Comment by problame** [comment]
@klauspost wow, that's cool. I'm a little short on time ATM due to my master's thesis.
AFAICT this is the plan for implementaiton:

- fit the `compress` option somewhere in the config of the sending side (maybe `send` options, but #285 already adds this field for ZFS's compressed send flag, so we'd need a more expressive name)
- extend the `SendReq` in `replication/logic/pdu/pdu.proto` with a field that specifies how the stream should be decompressed
 #285 will bring compressed 
  - I think an enum or a `oneof` would be best. Not sure how we'd deal with compressors that have different levels (gzip, zstd).
- Put the compressor somewhere into the send pipeline in 
  https://github.com/zrepl/zrepl/blob/180c3d9ae1a6e30078d0bdfd58ee73848a0dee01/endpoint%2Fendpoint.go#L143
- Put the decompressor somewhere in the recv pipeline in
  https://github.com/zrepl/zrepl/blob/180c3d9ae1a6e30078d0bdfd58ee73848a0dee01/endpoint%2Fendpoint.go#L634 
  - needs error handling in case the decompressor is not supported
- bump the protocol version
- testing ... I'd be fine with basic tests for the config that
  -  ensure the default value for the config flag is `off`
  - unsupported values produce an error

I won't get to this anytime soon but if anyone wants to pick this up I'm happy to do code review or answer questions about zrepl src "architecture". Start a WIP PR as you start working on it and mention it in this issue so people have something they can follow.

**Comment by beren12** [comment]
One thing about this, talking with the ppl in #zfsonlinux: -c won't recompress data ever, even if you specify -o compression=zstd or have some other compression algo set on the receiving dataset

**Comment by problame** [comment]
I think that if `-c` is not specified on the sender then the receiver's transforms apply (compression, encryption). But I could be wrong. Regardless, this issue is not about supporting -c, that's what #285 does.

**Comment by problame** [comment]
Oh, sorry, I think I lost track here, the issue was indeed about supporting `-c`. So that will land as part of #285.
@klauspost 's [comment above](https://github.com/zrepl/zrepl/issues/182#issuecomment-773984932) was about transport-level compression.

**Comment by problame** [comment]
@beren12 see [this issue](https://github.com/zrepl/zrepl/issues/197) for recv-side compression.

**Comment by problame** [comment]
I created https://github.com/zrepl/zrepl/issues/422 for the transport-level compression use case.
@all please subscribe to that issue to follow progress on transport-level compression. This issue will be closed by #285

**Comment by InsanePrawn** [comment]
Closed by #285 indeed.
